### PR TITLE
raftstore: queue vote

### PR DESCRIPTION
**Pull request by BusyJay**
If leader starts campaign after split immediately, follower may not apply the split yet. But if the follower is ready to apply the split, leader should probably apply the split already, so it should be more efficient to let follower start campaign than leader.

ref #1639 

@siddontang @zhangjinpeng1987 @hhkbp2 PTAL

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -1062,12 +1062,20 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
                 let is_leader = self.region_peers[&region_id].is_leader();
-                if is_leader && right.get_peers().len() > 1 {
+                let my_id = new_peer.peer_id();
+                if right.get_peers().into_iter().all(|p| p.get_id() >= my_id) {
+                    if right.get_peers().len() > 1 {
+                        new_peer.raft_group.raft.term -= 1;
+                        // Use a smaller term so the assigned leader won't have lease.
+                        new_peer.mut_store().applied_index_term -= 1;
+                        new_peer.raft_group.raft.become_candidate();
+                        new_peer.raft_group.raft.become_leader();
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: I think we must keep the same way with Raft paper, for a term if there is a Leader, other followers must all vote for it already.

**Comment by BusyJay** [comment]
PTAL

**Comment by siddontang** [comment]
I prefer using prevote here.

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -901,6 +901,56 @@ impl Peer {
...
+            // The peer campaigned when it was created, no need to do it again.
+            return;
+        }
+
+        if last_peer.is_leader() {
+            // If leader campaigns, follower may not apply yet.
+            return;
+        }
+
+        if self.region().get_peers().len() == 3 {
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: if the replicas number is 5, should we use 5 here?

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -586,9 +589,10 @@ impl<T: Transport, C: PdClient> Store<T, C> {
             .next() {
             let exist_region = self.region_peers[&exist_region_id].region();
             if enc_start_key(exist_region) < data_end_key(msg.get_end_key()) {
-                debug!("msg {:?} is overlapped with region {:?}, ignored",
+                debug!("msg {:?} is overlapped with region {:?}, queue for later split",
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: In my previous test, the new split leader vote messages will be dropped here https://github.com/pingcap/tikv/blob/48d44fdd861a94f7aed0774f85a6645b046f7fbd/src/raftstore/store/store.rs#L580

File: src/util/mod.rs [hunk 4]
```diff
@@ -420,6 +422,36 @@ pub fn monitor_threads<S: Into<String>>(_: S) -> io::Result<()> {
     Ok(())
 }
 
+/// A simple ring queue with fixed capacity.
```
Comment (ngaut) [inline src/util/mod.rs]: I guess there are some implementations. Please use a crate instead.

File: src/raftstore/store/store.rs (hunk 3 above)
Comment (BusyJay) [inline src/raftstore/store/store.rs]: How? It can't even pass the if condition check.

File: tests/raftstore/test_split_region.rs [hunk 5]
```diff
@@ -554,10 +554,11 @@ fn test_node_split_stale_epoch() {
 // it would accelerate ticks for the peer of new region after split, so that the peer of new region
 // could start campaign faster. and then this peer may take the leadership earlier.
 // `test_tick_acceleration_after_split` is a helper function for testing this feature.
```
Comment (hhkbp2) [inline tests/raftstore/test_split_region.rs]: Please update comments along with the code.

File: src/raftstore/store/store.rs [hunk 6]
```diff
@@ -1061,17 +1071,13 @@ impl<T: Transport, C: PdClient> Store<T, C> {
                 // `MsgRequestVote` from this peer on its campaign.
```
Comment (hhkbp2) [inline src/raftstore/store/store.rs]: Update comments.

File: src/raftstore/store/store.rs [hunk 7]
```diff
@@ -184,6 +186,7 @@ impl<T, C> Store<T, C> {
             peer_cache: Rc::new(RefCell::new(peer_cache)),
             snap_mgr: mgr,
             raft_metrics: RaftMetrics::default(),
+            pending_votes: FixRingQueue::with_capacity(10),
```
Comment (hhkbp2) [inline src/raftstore/store/store.rs]: Define a constant for 10.

**Comment by hhkbp2** [comment]
Any data on performance improvement of this PR?

**Comment by BusyJay** [comment]
This pr tries to resolve #1675. The policy has changed several times, I will clean the pr after it's verified to work.

File: src/raftstore/store/store.rs (hunk 3 above)
Comment (siddontang) [inline src/raftstore/store/store.rs]: sorry, I am wrong, I meet the condition here https://github.com/pingcap/tikv/blob/48d44fdd861a94f7aed0774f85a6645b046f7fbd/src/raftstore/store/store.rs#L717

```
2017/03/09 17:46:49.962 store.rs:650: [INFO] [region 24] handle raft message MsgRequestVote, from 25 to 26
2017/03/09 17:46:49.962 store.rs:1943: [ERROR] [store 3] handle raft message err: Other(StringError("[region 24] region not exist but not tombstone: state: Normal region {id: 24 start_key: \"t\\200\\000\\000\\000\\000\\000\\000\\377\\023_r\\200\\000\\000\\000\\000\\377\\0004\\370\\000\\000\\000\\000\\000\\372\" end_key: \"\" region_epoch {conf_ver: 3 version: 5} peers {id: 25 store_id: 1} peers {id: 26 store_id: 3} peers {id: 27 store_id: 2}}"))
```

Comment (BusyJay) [inline src/raftstore/store/store.rs]: It's handled properly.

File: src/util/mod.rs [hunk 8]
```diff
@@ -420,6 +422,36 @@ pub fn monitor_threads<S: Into<String>>(_: S) -> io::Result<()> {
...
+pub struct FixRingQueue<T> {
```
Comment (siddontang) [inline src/util/mod.rs]: Add test to cover the queue.

File: src/raftstore/store/store.rs [hunk 9]
```diff
@@ -1091,6 +1097,16 @@ impl<T: Transport, C: PdClient> Store<T, C> {
                 self.region_peers.insert(new_region_id, new_peer);
             }
         }
+
+        if !campaigned {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: seem `if campaigned` ?

File: src/util/mod.rs (hunk 4 above)
Comment (BusyJay) [inline src/util/mod.rs]: There are two implementations out there. They are either too complicated or has strange abstraction. The struct is very simple, no need to introduce another dependency.

**Comment by siddontang** [comment]
I prefer RingBuffer too.

Rest LGTM

**Comment by siddontang** [comment]
PTAL @hhkbp2 

LGTM

**Comment by siddontang** [comment]
Ping @hhkbp2

**Comment by hhkbp2** [comment]
LGTM

**Pull request merged**